---
title: "RNAseq Analysis with kallisto/Chris Deitrick, pt. 3"
output: html_notebook
---
```{r}
library(dplyr)
library(tidyverse)
library(DESeq2)
library(knitr)
library(apeglm)
library(here)
library(pcaExplorer)
library(genefilter)
library(pheatmap)
library(RColorBrewer)
library(viridis)
library(GenomicFeatures)
source(here::here("scripts", "deseq_workflow_expanded_mk3.r"), local=TRUE, echo=FALSE)
```

# Setup again

A long string of notebooks. Previously, I had found that ribosomal RNA reads (86 features) were missing from the annotation tables, so I  shuffled the original output into the "old" folder to save.

## Rerun kallisto with expanded index
The file I used to make the kallisto index did not include the rna transcripts, but they were included another file. So, I have appended both the _cds_from_genomic.fna and _rna_from_genomic.fna files together in a text editor, and saved as _cds_and_rna_from_genomic_unordered.fna. (Pushed them to my Beagle folder: /home/mjf123/BcenHI2424/TilS/TilS_RNA/data/indeces/)
These are for the RefSeq versions. (There are GenBank ones, too.)

On Beagle, in the ~/mjf123/BcenHI2424/TilS/TilS_RNA/data/indeces/ folder:
```{linux eval=FALSE}
module load kallisto
kallisto index --index=HI2424.RS.cds.index GCA_000203955.1_ASM20395v1_cds_and_rna_from_genomic_unordered.fna
sh ../../scripts/TilSkallisto.sh
```


## Making a new feature table

As for the annotation table, GenBank has a cleaner version than what I had patched together listed as a tab-delineated _feature_table.txt. Every feature appears to be listed twice, once as a gene and then again as cds/RNA. The second listing has most of the information, except for the RNA feature that are missing an entry under "class." (But not all entries are actually doubled, splitting them by even/odd indexes does not produce matching lists of locus_tags)

parse_GB_annotation filters out the duplicate information, writes a .tsv file for the run_on_kallisto.py script to use, and returns the annotation table.

```{r}

featurefile <- here::here("data", "references", "GCF_000203955.1_ASM20395v1_feature_table.txt")
anno_table <- parse_RS_annotation(featurefile) # For RefSeq, use parse_GB_annotation for GenBank
nrow(anno_table)
```

## Collect kallisto data into abundance table
Made new abundance tables using Chris Deitrick's "run_on_kallisto.py" script, giving the new annotation table. (Moved old abundance tables into old_tables folder.)
On beagle, from ~/BcenHI2424/TilS/TilS_RNA/:
```{python eval=FALSE}
python scripts/run_on_kallisto.py --reference data/indeces/GCF_000203955.1_ASM20395v1_cds_and_rna_from_genomic_unordered.fna --folder results/results_kallisto/ --filename results/abundance.all.tsv --annotations data/indeces/RS.annotations.tsv
```

## Subsetting the data
And downloaded the three abundance tables. (These DESeq2 scripts only use the .matrix.tsv and .design.tsv)

```{r}
allMatrix <- read_table_counts(here::here("data", "abundance.all.matrix.tsv"))
nrow(allMatrix)
allDesign <- read_table_design(here::here("data", "abundance.all.design.tsv"))
locusKey <- row.names(anno_table)
names(locusKey) <- anno_table$RSlocusTag

for ( listitem in names(locusKey)){
  if (listitem %in% rownames(allMatrix)){
     rownames(allMatrix)[ rownames(allMatrix) == listitem] <- locusKey[[listitem]]
  } 
}

annonames <- rownames(anno_table)
matrixnames <- rownames(allMatrix)
missingnames <- annonames[!(annonames %in% matrixnames)]
print(missingnames)
```
As a note, the counts table does not have rows for the 26 pseudo genes. This may be because the pseudo genes aren't in the fasta files used to run kallisto. *Not sure how to include these yet.*

Subetting first by type in the design tables, then using the design table names to subset the matrices.
```{r}
tilSDesign <- allDesign[(allDesign$type %in% "tils"), ]
tilSMatrix <- allMatrix[ , rownames(tilSDesign)]
ppcDesign <- allDesign[(allDesign$type %in% "ppc") | (allDesign$strain == "WT"), ]
ppcMatrix <- allMatrix[ , rownames(ppcDesign)]
```

# Rerunning DESeq2 with new annotation tables
Now, re-run everthing else....

```{r}
# Full data set: with and without chromosome 3
desOutFull <- des_workflow("results/DESeq2_Results_Full", allMatrix, "plmd", allDesign, anno_table)
doFull <- des_workflow("results/DESeq2_Results_Full_noChrom3", allMatrix, "both", allDesign , anno_table)

# TilS subset: with and without chromosome 3
desOutTilS <- des_workflow("results/DESeq2_Results_TilS", tilSMatrix, "plmd", tilSDesign, anno_table)
doTilS <- des_workflow("results/DESeq2_Results_TilS_noChrom3", tilSMatrix, "both", tilSDesign , anno_table)

# PPC subset: with and without chromosome 3
desOutPPC <- des_workflow("results/DESeq2_Results_PPC", ppcMatrix, "plmd", ppcDesign, anno_table)
doPPC <- des_workflow("results/DESeq2_Results_PPC_noChrom3", ppcMatrix, "both", ppcDesign , anno_table)

dds <- list(desOutFull[[1]], doFull[[1]], desOutTilS[[1]], doTilS[[1]], desOutPPC[[1]], doPPC[[1]])
desres <- list(desOutFull[[2]], doFull[[2]], desOutTilS[[2]], doTilS[[2]], desOutPPC[[2]], doPPC[[2]])
nameKey <- c("Full", "Full_nochr3", "TilS", "TilS_nochr3", "PPC", "PPC_nochr3")
names(dds) <- nameKey
names(desres) <- nameKey
rm(desOutFull, desOutTilS, desOutPPC, doFull, doTilS, doPPC) # Clean up
```

## Filtering and collating pairwise results
Using padj < 0.1 as the cutoff. Create list of data frames with signficant genes only for each pairing, from within the list of datasets. (Adding annotation description column, too.)
(Old way gave wrong "name" to genes.)
```{r}
desresSig <- lapply(desres, function(matrixof) {
  lapply(matrixof[ ,1], function(df) {
    newdf <- df[order(df$pvalue), ]
    newdf <- subset(newdf, padj<0.1)
    newdf <- as.data.frame(newdf)
    # newdf <- tibble::rownames_to_column(newdf, var="Gene")
    
    newanno <- vector("character", nrow(newdf))
    names(newanno) <- rownames(newdf)
    for (listitem in rownames(newdf)){
      newanno[listitem] <- anno_table[listitem, "name"]
    }
    newdf <- merge(newdf, as.data.frame(newanno), by="row.names", all=TRUE)
    return (newdf)
  })
})
```

Collating pairwise results for each dataset, and writing to csv file in main results folder. Each gene can be listed multiple times, but will have column entries denoting with mutant compared to which mutant.
```{r}

desresCol <- lapply(nameKey, function(listitem){
  wordList <- as.list(names(desresSig[[listitem]]))
  newdf <- mapply( function(df, w ) {words <- unlist(strsplit(w, "_to_", fixed=TRUE))
    df <- cbind( df, "mutant"=rep(words[[1]], nrow(df)),
                        "comparedTo"=rep(words[[2]], nrow(df)))
    }, desresSig[[listitem]], wordList, SIMPLIFY = FALSE, USE.NAMES = FALSE)
  collated <- do.call(rbind, newdf)
  newfile <- here::here("results", paste0("collated_sigRes_",listitem,".csv"))
  write.csv(collated, file=newfile, row.names = FALSE)
  return (collated)
})
```


# Rlog transformation
Using the rlog transformations, as the data set is less than 30. ([For reasoning](https://bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html#the-variance-stabilizing-transformation-and-the-rlog))
Chose a non-blind transformation, because we expect to see differences dependent on mutant. See also a discussion on the [blind parameter.](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#blind-dispersion-estimation)

The DESeq2Transform objects produced can be given to pcaExplorer as well.
```{r}
rld <- lapply(setNames(nameKey, nameKey), function(listitem){
  rlog(dds[[listitem]], blind=FALSE)
})
```

## Heatmaps of the transformed count matrix
Creating heatmaps from each of the rlog transformations.
```{r}
selector <- vector("list", 8)
selector <- lapply(setNames(nameKey,nameKey), function(listitem){
  order(rowMeans(counts(dds[[listitem]], normalized=TRUE)), decreasing=TRUE)[1:20]
})

titles <- c("Full Data Set", "Full Data Set, without chromosome 3",
            "TilS Subset", "TilS Subset, without chromosome 3",
            "PPC Subset", "PPC Subset, without chromosome 3")
names(titles) <- nameKey
df <- as.data.frame(colData(dds[[1]])[,c("strain","type")]) #this is just a key

dir.create(here::here("results", "count_heatmaps"), showWarnings = FALSE)
heatmaps <- lapply(setNames(nameKey,nameKey), function(listitem){
  pheatmap(assay(rld[[listitem]])[selector[[listitem]],], cluster_rows=FALSE, show_rownames=FALSE,
           cluster_cols=FALSE, annoation_col=df, color = magma(255, begin = 0.15),
           main = paste0("Heatmap of Count Matrix for ", titles[[listitem]]),
           filename = here::here("results", "count_heatmaps", paste0("rld_heatmap_", listitem, ".pdf")))
})
```

## Clustering
With the transformations, create clusterings based on distances, and save plots.

```{r echo=FALSE}
colors <- colorRampPalette( rev(brewer.pal(9, "PuBuGn")) )(255)

sampleDists <- vector("list", 8)
sampleDists <- lapply(setNames(nameKey,nameKey), function(listitem){
  sampleDists[[listitem]] <- dist(t(assay(rld[[listitem]])))
})

```

Plot and save all of the clusterings in one file. (To create separate files, comment out the pdf and dev lines here, and then uncomment the same lines within the function in the deseq_workflow_expanded_mk3.r script. Creating a funciton for later use fixed problems with the lapply call.)
```{r}
dir.create(here::here("results", "rld_clustering"), showWarnings = FALSE)
# pdf(file=here::here("results", "rld_clustering", "rldclusterings_All.pdf"))
rldclusters <- lapply(nameKey, function(listitem){
  fname <- here::here("results", "rld_clustering", paste0("rldclustering_", listitem, ".pdf"))
  clustering_plot(sampleDists[[listitem]], rld[[listitem]], colors, titles[[listitem]], fname)
})
# dev.off()
```


## Clustering by top variance (not using log2 differential)

Creating lists of highest variance genes (cutoff of 20), comparing with the average across all of the counts for each gene within the rlog transformed data set in order to cluster. Because this takes into account variance, PPC and TilS should definitely be separated as the PPC counts are considerably lower. saving tables and plots.
(Old way gave wrong "name" to genes.)
```{r}

topVarGenes <- lapply(setNames(nameKey,nameKey), function(listitem){
  head(order(rowVars(assay(rld[[listitem]])), decreasing = TRUE), 50)
})
topVarEntries <- vector("list", length(nameKey))
topVarEntries <- lapply(setNames(nameKey, nameKey), function(listitem){
  topVarEntries[[listitem]]  <- assay(rld[[listitem]])[ topVarGenes[[listitem]], ]
})
dir.create(here::here("results", "topGenes_byVariance_acrossDataset"), showWarnings = FALSE)
topVarEntriesAno <- vector("list", length(nameKey))
topVarEntriesAno <- lapply(setNames(nameKey, nameKey), function(listitem){

  topVarEntriesAno[[listitem]] <- as.data.frame(topVarEntries[[listitem]])
  newanno <- vector("character", nrow(topVarEntriesAno[[listitem]]))
  names(newanno) <- rownames(topVarEntriesAno[[listitem]])
    
  for (entries in rownames(topVarEntries[[listitem]])){
    newanno[entries] <- anno_table[entries, "name"]
  }
  topVarEntriesAno[[listitem]] <- merge(topVarEntriesAno[[listitem]], as.data.frame(newanno), by="row.names", all=TRUE)

  fname <- here::here("results", "topGenes_byVariance_acrossDataset", paste0("TopVarianceGenes_",listitem,".csv"))
  write.csv(topVarEntriesAno[[listitem]], file=fname, row.names=FALSE)
})
```

```{r}
pheat <- lapply(setNames(nameKey, nameKey), function(listitem){
  mat  <- topVarEntries[[listitem]] - rowMeans(topVarEntries[[listitem]])
  anno <- as.data.frame(colData(rld[[listitem]])[, c("strain","type")])
  fname <- here::here("results", "topGenes_byVAriance_acrossDataset", paste0("geneVarCluster_", listitem, ".pdf"))
  pheatmap(mat, annotation_col = anno, fontsize_number = 2,
           cellwidth = 12, cellheight = 9,
           main=paste0("Clustering of Highest Variance Genes for ", titles[[listitem]]),
           filename = fname)
})
```
 
*Other clustering schema?*

# PCA plots

Set a color scheme so that each mutant stays the same color.
```{r}
mutantColors <- brewer.pal(8, "Set1")
names(mutantColors) <- levels(allDesign$strain)
```

Create, save, and store pca plots. (dimentions could be defined in the ggsave call)
```{r}
dir.create(here::here("results","pca"), showWarnings = FALSE)
pcas <- lapply(setNames(nameKey,nameKey), function(listitem) {
  pcaData <- plotPCA(rld[[listitem]], intgroup=c("strain", "type"), returnData=TRUE)
  percentVar <- round(100 * attr(pcaData, "percentVar"))
  fname <- here::here("results", "pca", paste0("pcaPlot_", listitem, ".pdf"))
  p <- ggplot(pcaData, aes(PC1, PC2, color=strain, shape=type)) +
    geom_point(size=3) +
    xlab(paste0("PC1: ",percentVar[1],"% variance")) +
    ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
    coord_fixed() +
    ggtitle(paste0("Principle Component Analysis, ",listitem)) +
    theme(plot.title = element_text(hjust = 0.5)) +
    scale_color_manual(values = mutantColors)
  ggsave(fname, device = "pdf")
  plot(p)
})
```

## PCA EXplorer
The interacitve viewer is useful for grabbing the top and bottom loading genes for each principle component. Adding a column to add the gene description to the locus tag.
```{r}
anno_table$gene_name <- paste(anno_table$locusTag, anno_table$symbol, anno_table$name, sep = "-")
```

Running pcaExplorer on each data set. Mostly interested in the TilS and PPC subsets. Saving the top/bottom loading genes, expression projections, and the variance explained by PC (scree).
(I don't think that I need to use any of the options on the first tab since I'm uploading already transformed data.)
Also, trying to save more than 15 top/bottom loading genes in the sample tab loses some text.
Settings: Let nr of most variant gene be 300.
The loadings plots need to be spaced out better in Inkscape. (I think that those loadings are the same as the top variance genes calculated earlier. But would like to verify and look to plot the CI ovals manually for more control?)
Also able to to see what it looks like to leave out N445K3 and tRNA2 as possible outliers. Groupings don't change too much.


```{r}
pcaExplorer(dds = dds[["TilS"]], dst = rld[["TilS"]], coldata = tilSDesign, annotation = anno_table)
```

```{r}
pcaExplorer(dds = dds[["TilS_nochr3"]], dst = rld[["TilS_nochr3"]], coldata = tilSDesign, annotation = anno_table)
```
```{r}
pcaExplorer(dds = dds[["PPC"]], dst = rld[["PPC"]], coldata = ppcDesign, annotation = anno_table)
```

```{r}
pcaExplorer(dds = dds[["PPC_nochr3"]], dst = rld[["PPC_nochr3"]], coldata = ppcDesign, annotation = anno_table)
```

```{r}
pcaExplorer(dds = dds[["Full"]], dst = rld[["Full"]], coldata = allDesign, annotation = anno_table)
```
```{r}
pcaExplorer(dds = dds[["Full_nochr3"]], dst = rld[["Full_nochr3"]], coldata = allDesign, annotation = anno_table)
```
Table export options are down at the bottom of the genes panel: look at tightly clustered genes? (Full data set is actually interesting to see in that window.)


# Gviz?
*Current goals:*

*Update pcaExplorer figures, re-sizing and formatting the pcaExplorer top/bottom loading figures
*Attaching the GRanges data to the DESEqDataSets.
  +Do I need rtracklayer?
*gviz up and running, requires pairwise comparisons... and grange, and lfcShrink
*Back through pcaExplorer, look at gene tab

(remove tRNA mutant from TilS set and see how things change? The ribosomal RNA appears to be dominating results. Should I remove only the ribosomal?)

```{r}
library(GenomicFeatures)
txdb <- makeTxDbFromGFF(here::here("data", "references", "GCA_000203955.1_ASM20395v1_genomic.gff"))
txdb
txdbRS <- makeTxDbFromGFF(here::here("data", "references", "GCF_000203955.1_ASM20395v1_genomic.gff")) # RefSeq?
txdbRS
```
It looks like the GenBank version doesn't include the RNA features, but the RefSeq does. cds nrow is different, too. RefSeq dropped: gene-BCEN2424_RS16455, gene-BCEN2424_RS17110, gene-BCEN2424_RS33785, gene-BCEN2424_RS34535OK


```{r}
sessionInfo()
```

