---
title: "RNAseq Analysis with kallisto/Chris Deitrick, pt. 3"
output: html_notebook
---
```{r}
library(plyr)
library(dplyr)
library(tidyverse)
library(DESeq2)
library(knitr)
library(apeglm)
library(pcaExplorer)
library(genefilter)
library(pheatmap)
library(RColorBrewer)
library(viridis)
library(GenomicFeatures)
library(ggpubr)
library(ggrepel)
library(here)
source(here::here("scripts", "deseq_workflow_expanded_mk3.r"), local=TRUE, echo=FALSE)
```
Also the two DESeq2 vignettes, [one published with its paper](https://bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html#ma-plot) and [one being updated post-paper with extra tidbits](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#more-information-on-results-columns), and [this DESeq2 walkthrough](https://hbctraining.github.io/DGE_workshop/lessons/05_DGE_DESeq2_analysis2.html) were very helpful.

# Setup again

A long string of notebooks. Previously, I had found that ribosomal RNA reads (86 features) were missing from the annotation tables, so I  shuffled the original output into the "old" folder to save.

## Rerun kallisto with expanded index
The file I used to make the kallisto index did not include the rna transcripts, but they were included another file. So, I have appended both the _cds_from_genomic.fna and _rna_from_genomic.fna files together in a text editor, and saved as _cds_and_rna_from_genomic_unordered.fna. (Pushed them to my Beagle folder: /home/mjf123/BcenHI2424/TilS/TilS_RNA/data/indeces/)
These are for the RefSeq versions. (There are GenBank ones, too.)

On Beagle, in the ~/mjf123/BcenHI2424/TilS/TilS_RNA/data/indeces/ folder:
```{linux eval=FALSE}
module load kallisto
kallisto index --index=HI2424.RS.cds.index GCA_000203955.1_ASM20395v1_cds_and_rna_from_genomic_unordered.fna
sh ../../scripts/TilSkallisto.sh
```


## Making a new feature table

As for the annotation table, GenBank has a cleaner version than what I had patched together listed as a tab-delineated _feature_table.txt. Every feature appears to be listed twice, once as a gene and then again as cds/RNA. The second listing has most of the information, except for the RNA feature that are missing an entry under "class." (But not all entries are actually doubled, splitting them by even/odd indexes does not produce matching lists of locus_tags)

parse_GB_annotation filters out the duplicate information, writes a .tsv file for the run_on_kallisto.py script to use, and returns the annotation table.

```{r}

featurefile <- here::here("data", "references", "GCF_000203955.1_ASM20395v1_feature_table.txt")
anno_table <- parse_RS_annotation(featurefile) # For RefSeq, use parse_GB_annotation for GenBank
nrow(anno_table)
```

## Collect kallisto data into abundance table
Made new abundance tables using Chris Deitrick's "run_on_kallisto.py" script, giving the new annotation table. (Moved old abundance tables into old_tables folder.)
On beagle, from ~/BcenHI2424/TilS/TilS_RNA/:
```{python eval=FALSE}
python scripts/run_on_kallisto.py --reference data/indeces/GCF_000203955.1_ASM20395v1_cds_and_rna_from_genomic_unordered.fna --folder results/results_kallisto/ --filename results/abundance.all.tsv --annotations data/indeces/RS.annotations.tsv
```

## Subsetting the data
Downloaded the three abundance tables from Beagle. (These DESeq2 scripts only use the .matrix.tsv and .design.tsv) Changed the names to  match GB naming. Then ordered the counts and annotation tables to help filtering.

```{r}
allMatrix <- read_table_counts(here::here("data", "abundance.all.matrix.tsv"))
nrow(allMatrix)
allDesign <- read_table_design(here::here("data", "abundance.all.design.tsv"))

#Change names
locusKey <- row.names(anno_table)
names(locusKey) <- anno_table$RSlocusTag
for ( listitem in names(locusKey)){
  if (listitem %in% rownames(allMatrix)){
      rownames(allMatrix)[ rownames(allMatrix) == listitem] <- locusKey[[listitem]]
  } 
}
allMatrix <- allMatrix[ order(row.names(allMatrix)), ] #order the count table
anno_table <- anno_table[ order(row.names(anno_table)), ]
```
As a note, the counts table does not have rows for the 26 pseudo genes. This may be because the pseudo genes aren't in the fasta files used to run kallisto. *Not sure how to include these yet.*

Subetting first by type in the design tables, then using the design table names to subset the matrices.
```{r}
tilSDesign <- allDesign[(allDesign$type %in% "tils"), ]
tilSMatrix <- allMatrix[ , rownames(tilSDesign)]
ppcDesign <- allDesign[(allDesign$type %in% "ppc") | (allDesign$strain == "WT"), ]
ppcMatrix <- allMatrix[ , rownames(ppcDesign)]
```

# Rerunning DESeq2 with new annotation tables
Now, re-run everthing else....
(des_workflow includes switching to GB naming for the DESEq2DataSet and results. The allMatrix is still in RS.)
```{r}
# Full data set: with and without chromosome 3
desOutFull <- des_workflow("results/DESeq2_Results_Full", allMatrix, "plmd", allDesign, anno_table)
doFull <- des_workflow("results/DESeq2_Results_Full_noChrom3", allMatrix, "both", allDesign , anno_table)

# TilS subset: with and without chromosome 3
desOutTilS <- des_workflow("results/DESeq2_Results_TilS", tilSMatrix, "plmd", tilSDesign, anno_table)
doTilS <- des_workflow("results/DESeq2_Results_TilS_noChrom3", tilSMatrix, "both", tilSDesign , anno_table)

# PPC subset: with and without chromosome 3
desOutPPC <- des_workflow("results/DESeq2_Results_PPC", ppcMatrix, "plmd", ppcDesign, anno_table)
doPPC <- des_workflow("results/DESeq2_Results_PPC_noChrom3", ppcMatrix, "both", ppcDesign , anno_table)

dds <- list(desOutFull[[1]], doFull[[1]], desOutTilS[[1]], doTilS[[1]], desOutPPC[[1]], doPPC[[1]])
desres <- list(desOutFull[[2]], doFull[[2]], desOutTilS[[2]], doTilS[[2]], desOutPPC[[2]], doPPC[[2]])
nameKey <- c("Full", "Full_nochr3", "TilS", "TilS_nochr3", "PPC", "PPC_nochr3")
names(dds) <- nameKey
names(desres) <- nameKey
rm(desOutFull, desOutTilS, desOutPPC, doFull, doTilS, doPPC) # Clean up
```
Checking that filtering actually worked. (Only looking at the biggest data sets.)
```{r}
inP <- as.character(anno_table[ anno_table$chromosome == "Plasmid1", "locusTag"])
in3 <- as.character( anno_table[ anno_table$chromosome == "3", "locusTag"] )
any( rownames(dds[[1]]) %in% inP )
any( rownames(dds[[2]]) %in% c(inP, in3))
```

### Null entires
There are some null entires in the comparisons to WT. See MA-plots section for exact genes.


## Filtering and collating pairwise results
Using padj < 0.1 as the cutoff. Create list of data frames with signficant genes only for each pairing, from within the list of datasets. (Adding annotation description column, too.)
(Old way gave wrong "name" to genes.)
```{r}
desresSig <- lapply(desres, function(matrixof) {
  lapply(matrixof[ ,1], function(df) {
    newdf <- df[order(df$pvalue), ]
    newdf <- subset(newdf, padj<0.1)
    newdf <- as.data.frame(newdf)
    newdf <- tibble::rownames_to_column(newdf, var="locusTag") #to use join_all later
    newdf$abs_log2 <- abs(newdf$log2FoldChange) # Add absolute value columns
    newdf$abs_stat <- abs(newdf$stat)
    newanno <- anno_table[(anno_table$locusTag %in% newdf$locusTag), ] %>%
      dplyr::select("locusTag", "name", "chromosome") # Subset anno_table to match the shorter list
    newdf <- join_all( list(newdf, newanno), by = "locusTag", type = "full" )
    return (newdf)
  })
})
```
*(interesting to note which pairwise did not return any results meeting the cutoff.)*
Collating pairwise results for each dataset, and writing to csv file in main results folder. Each gene can be listed multiple times, but will have column entries denoting with mutant compared to which mutant.
```{r}
dir.create(here::here("results", "pairwise"), showWarnings = FALSE)
desresCol <- lapply(nameKey, function(listitem){
  wordList <- as.list(names(desresSig[[listitem]]))
  newdf <- mapply( function(df, w ) {words <- unlist(strsplit(w, "_to_", fixed=TRUE))
    df <- cbind( df, "mutant"=rep(words[[1]], nrow(df)),
                        "comparedTo"=rep(words[[2]], nrow(df)))
    }, desresSig[[listitem]], wordList, SIMPLIFY = FALSE, USE.NAMES = FALSE)
  collated <- do.call(rbind, newdf)
  newfile <- here::here("results", "pairwise", paste0("collated_sigRes_",listitem,".csv"))
  write.csv(collated, file=newfile, row.names = FALSE)
  return (collated)
})
names(desresCol) <- nameKey
```
(No N445K results made it past the filtering.)

# Rlog transformation
Using the rlog transformations, as the data set is less than 30. ([For reasoning](https://bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html#the-variance-stabilizing-transformation-and-the-rlog))
Chose a non-blind transformation, because we expect to see differences dependent on mutant. See also a discussion on the [blind parameter.](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#blind-dispersion-estimation)

The DESeq2Transform objects produced can be given to pcaExplorer as well.
```{r}
rld <- lapply(setNames(nameKey, nameKey), function(listitem){
  rlog(dds[[listitem]], blind=FALSE)
})
```

## Heatmaps of the transformed count matrix
Creating heatmaps from each of the rlog transformations.
```{r}
selector <- vector("list", 8)
selector <- lapply(setNames(nameKey,nameKey), function(listitem){
  order(rowMeans(counts(dds[[listitem]], normalized=TRUE)), decreasing=TRUE)[1:20]
})

titles <- c("Full Data Set", "Full Data Set, without chromosome 3",
            "TilS Subset", "TilS Subset, without chromosome 3",
            "PPC Subset", "PPC Subset, without chromosome 3")
names(titles) <- nameKey
df <- as.data.frame(colData(dds[[1]])[,c("strain","type")]) #this is just a key

dir.create(here::here("results", "count_heatmaps"), showWarnings = FALSE)
heatmaps <- lapply(setNames(nameKey,nameKey), function(listitem){
  pheatmap(assay(rld[[listitem]])[selector[[listitem]],], cluster_rows=FALSE, show_rownames=FALSE,
           cluster_cols=FALSE, annoation_col=df, color = magma(255, begin = 0.15),
           main = paste0("Heatmap of Count Matrix for ", titles[[listitem]]),
           filename = here::here("results", "count_heatmaps", paste0("rld_heatmap_", listitem, ".pdf")))
})
```

## Clustering
With the transformations, create clusterings based on distances, and save plots.

```{r echo=FALSE}
colors <- colorRampPalette( rev(brewer.pal(9, "PuBuGn")) )(255)

sampleDists <- vector("list", 8)
sampleDists <- lapply(setNames(nameKey,nameKey), function(listitem){
  sampleDists[[listitem]] <- dist(t(assay(rld[[listitem]])))
})

```

Plot and save all of the clusterings in one file. (To create separate files, comment out the pdf and dev lines here, and then uncomment the same lines within the function in the deseq_workflow_expanded_mk3.r script. Creating a funciton for later use fixed problems with the lapply call.)
```{r}
dir.create(here::here("results", "rld_clustering"), showWarnings = FALSE)
# pdf(file=here::here("results", "rld_clustering", "rldclusterings_All.pdf"))
rldclusters <- lapply(nameKey, function(listitem){
  fname <- here::here("results", "rld_clustering", paste0("rldclustering_", listitem, ".pdf"))
  clustering_plot(sampleDists[[listitem]], rld[[listitem]], colors, titles[[listitem]], fname)
})
# dev.off()
```


## Clustering by top variance (not using log2 differential)

Creating lists of highest variance genes (cutoff of 20), comparing with the average across all of the counts for each gene within the rlog transformed data set in order to cluster. Because this takes into account variance, PPC and TilS should definitely be separated as the PPC counts are considerably lower. saving tables and plots.
(Old way gave wrong "name" to genes.)
```{r}

topVarGenes <- lapply(setNames(nameKey,nameKey), function(listitem){
  head(order(rowVars(assay(rld[[listitem]])), decreasing = TRUE), 50)
})
topVarEntries <- vector("list", length(nameKey))
topVarEntries <- lapply(setNames(nameKey, nameKey), function(listitem){
  topVarEntries[[listitem]]  <- assay(rld[[listitem]])[ topVarGenes[[listitem]], ]
})
dir.create(here::here("results", "topGenes_byVariance_acrossDataset"), showWarnings = FALSE)
topVarEntriesAno <- vector("list", length(nameKey))
topVarEntriesAno <- lapply(setNames(nameKey, nameKey), function(listitem){

  topVarEntriesAno[[listitem]] <- as.data.frame(topVarEntries[[listitem]])
  newanno <- vector("character", nrow(topVarEntriesAno[[listitem]]))
  names(newanno) <- rownames(topVarEntriesAno[[listitem]])
    
  for (entries in rownames(topVarEntries[[listitem]])){
    newanno[entries] <- anno_table[entries, "name"]
  }
  topVarEntriesAno[[listitem]] <- merge(topVarEntriesAno[[listitem]], as.data.frame(newanno), by="row.names", all=TRUE)

  fname <- here::here("results", "topGenes_byVariance_acrossDataset", paste0("TopVarianceGenes_",listitem,".csv"))
  write.csv(topVarEntriesAno[[listitem]], file=fname, row.names=FALSE)
})
```

```{r}
pheat <- lapply(setNames(nameKey, nameKey), function(listitem){
  mat  <- topVarEntries[[listitem]] - rowMeans(topVarEntries[[listitem]])
  anno <- as.data.frame(colData(rld[[listitem]])[, c("strain","type")])
  fname <- here::here("results", "topGenes_byVAriance_acrossDataset", paste0("geneVarCluster_", listitem, ".pdf"))
  pheatmap(mat, annotation_col = anno, fontsize_number = 2,
           cellwidth = 12, cellheight = 9,
           main=paste0("Clustering of Highest Variance Genes for ", titles[[listitem]]),
           filename = fname)
})
```
 
*Other clustering schema?*

# PCA plots

Set a color scheme so that each mutant stays the same color.
```{r}
mutantColors <- brewer.pal(8, "Set1")
names(mutantColors) <- levels(allDesign$strain)
```

Create, save, and store pca plots. (dimentions could be defined in the ggsave call)
```{r}
dir.create(here::here("results","pca"), showWarnings = FALSE)
pcas <- lapply(setNames(nameKey,nameKey), function(listitem) {
  pcaData <- plotPCA(rld[[listitem]], intgroup=c("strain", "type"), returnData=TRUE)
  percentVar <- round(100 * attr(pcaData, "percentVar"))
  fname <- here::here("results", "pca", paste0("pcaPlot_", listitem, ".pdf"))
  p <- ggplot(pcaData, aes(PC1, PC2, color=strain, shape=type)) +
    geom_point(size=3) +
    xlab(paste0("PC1: ",percentVar[1],"% variance")) +
    ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
    coord_fixed() +
    ggtitle(paste0("Principle Component Analysis, ",listitem)) +
    theme(plot.title = element_text(hjust = 0.5)) +
    scale_color_manual(values = mutantColors)
  ggsave(fname, device = "pdf")
  plot(p)
})
```

## PCA EXplorer
The interacitve viewer is useful for grabbing the top and bottom loading genes for each principle component. Adding a column to add the gene description to the locus tag.
```{r}
anno_table$gene_name <- paste(anno_table$locusTag, anno_table$symbol, anno_table$name, sep = "-")
write.csv(anno_table, file=here::here("data", "references", "anno_table.csv"), row.names = FALSE)
```

Running pcaExplorer on each data set. Mostly interested in the TilS and PPC subsets. Saving the top/bottom loading genes, expression projections, and the variance explained by PC (scree).
(I don't think that I need to use any of the options on the first tab since I'm uploading already transformed data.)
Also, trying to save more than 15 top/bottom loading genes in the sample tab loses some text.
Settings: Let nr of most variant gene be 300.
The loadings plots need to be spaced out better in Inkscape. (I think that those loadings are the same as the top variance genes calculated earlier. But would like to verify and look to plot the CI ovals manually for more control?)
Also able to to see what it looks like to leave out N445K3 and tRNA2 as possible outliers. Groupings don't change too much.


```{r}
pcaExplorer(dds = dds[["TilS"]], dst = rld[["TilS"]], coldata = tilSDesign, annotation = anno_table)
```

```{r}
pcaExplorer(dds = dds[["TilS_nochr3"]], dst = rld[["TilS_nochr3"]], coldata = tilSDesign, annotation = anno_table)
```
```{r}
pcaExplorer(dds = dds[["PPC"]], dst = rld[["PPC"]], coldata = ppcDesign, annotation = anno_table)
```

```{r}
pcaExplorer(dds = dds[["PPC_nochr3"]], dst = rld[["PPC_nochr3"]], coldata = ppcDesign, annotation = anno_table)
```

```{r}
pcaExplorer(dds = dds[["Full"]], dst = rld[["Full"]], coldata = allDesign, annotation = anno_table)
```
```{r}
pcaExplorer(dds = dds[["Full_nochr3"]], dst = rld[["Full_nochr3"]], coldata = allDesign, annotation = anno_table)
```
Table export options are down at the bottom of the genes panel: look at tightly clustered genes? (Full data set is actually interesting to see in that window.)

# LFC and MA-Plots
For these, going to focus in on the WT-as-reference set of data from the results. Generating the list of matchups for each data set.
```{r}
matchups <- lapply(setNames(nameKey, nameKey), function(listitem){
  resultsNames(dds[[listitem]])
})
```

## LFC Shrinkage
Calculating the shrinkage of effect size for each WT pairing of each dataset.
(Could have also supplied the results object (desres) for each matchup, which replaces the coef and type parameters.)
```{r}
resLFCWT <- lapply(setNames(nameKey, nameKey), function(listitem){
  dir.create(here::here("results", paste0("LFCShrunk_results_", listitem)), showWarnings = FALSE)
  lapply(setNames(matchups[[listitem]], matchups[[listitem]]), function(eachmatch){
    if( as.character(eachmatch) != "Intercept"){
      fname=here::here("results", paste0("LFCShrunk_results_", listitem), paste0("LFCShrunk_", eachmatch, ".csv"))
      print(eachmatch)
      df <- lfcShrink(dds[[listitem]], coef = eachmatch, type = "apeglm")
      write.csv(df, file = fname, row.names = FALSE)
      return(df)
    }
  })
})
resLFCWT <- lapply(setNames(nameKey, nameKey), function(listitem){ compact(resLFCWT[[listitem]])}) #Remove blanks
```

Can further filter the results using the lfcShrink results (which attempt to weight by dispersion in the normalized counts for each gene (less dispersion is given higher weighting/closer to original log2foldchange.))
Then save a collated list for each data set.

```{r}
# lfcShurnk 0.58 corresponds to 1.5 lf2c
padj.cutoff <- 0.1
lfc.cutoff <- 0.58

sigresLFCWT <- lapply(setNames(nameKey, nameKey), function(listitem) {
  lapply(resLFCWT[[listitem]], function(df) {
    if ( !(dim(df)[1] == 0) ){
      newdf <- df %>%  data.frame() %>% rownames_to_column(var="locusTag") %>% as_tibble()
      newdf <- newdf %>% filter(padj < padj.cutoff & abs(log2FoldChange) > lfc.cutoff)
    
      newdf$abs_log2 <- abs(newdf$log2FoldChange) # Add absolute value column
      newanno <- anno_table[(anno_table$locusTag %in% newdf$locusTag), ] %>%
        dplyr::select("locusTag", "name", "chromosome") # Subset anno_table to match the shorter list
      newdf <- join_all( list(newdf, newanno), by = "locusTag", type = "full" )
    }
    return (newdf)
  })
})

sigresLFCWTCol <- lapply(nameKey, function(listitem){
  wordList <- names(sigresLFCWT[[listitem]])
  newdf <- mapply( function(df, w ) {
    w <- sub("strain_", "", w)
    words <- unlist(strsplit(w, "_vs_", fixed=TRUE))
    df <- cbind( df, "mutant"=rep(words[[1]], nrow(df)),
                        "comparedTo"=rep(words[[2]], nrow(df)))
    }, sigresLFCWT[[listitem]], wordList, SIMPLIFY = FALSE, USE.NAMES = FALSE)
  collated <- do.call(rbind, newdf)
  newfile <- here::here("results", "pairwise", paste0("sigShrunkRes_WTref_",listitem,".csv"))
  write.csv(collated, file=newfile, row.names = FALSE)
  return (collated)
})
names(sigresLFCWTCol) <- nameKey
```
Notably, *none* of the N445K survived the cutoff. The problem appears to be in the padj values (which are super high).
```{r}
any(resLFCWT[["TilS_nochr3"]][["strain_N445K_vs_WT"]]$padj < 0.35)
any(resLFCWT[["TilS_nochr3"]][["strain_N445K_vs_WT"]]$padj < 0.3)
```
*And tRNA only had 1 result in both data sets: Bcen2424_1362*

(lfcShrink can also output in GRanges format, which I what I think I need for GViz plots.)
## MA-Plots
Got a lot of help from the standard DESeq2 vignettes, but also [this tutorial.](https://hbctraining.github.io/DGE_workshop/lessons/05_DGE_DESeq2_analysis2.html)

Plotting each individually allows interactive caputring of outlier genes. (Also requires the corresponding desres object.)

The desres lists are somewhat difficult to navigate. For the full data sets (1 and 2) the WT references are in the sublists at multiples of 7 (6 total), ing TilS subsets at the multiples of 5 (4 total), and in PPC subsets at the multiples of 2 (2 of them.)
```{r}
print("Names of resLFCWT elements")
for(listitem in nameKey){
  print(listitem)
  print(names(resLFCWT[[listitem]]))
}

desresWTfull <- lapply( 1:2, function(i){
  lapply(1:7, function(j){ desres[[i]][[j*7]] })
})
desresWTTilS <- lapply( 3:4, function(i){
  lapply(1:5, function(j){ desres[[i]][[j*5]] })
})
desresWTPPC <- lapply( 5:6, function(i){
  lapply(1:2, function(j){ desres[[i]][[j*2]] })
})
desresWT <- c(desresWTfull, desresWTTilS, desresWTPPC)
rm(desresWTfull, desresWTTilS, desresWTPPC)
#Name the elements for easier reference by mutant being compared
names(desresWT) <- nameKey
mutnameKey<- lapply(setNames(nameKey, nameKey), function(listitem){
   muts <- as.character(levels(dds[[listitem]]$strain))
   muts <- muts[muts != "WT"]
 })
print("Names of desresWT elements")
print(mutnameKey)
for (listitem in nameKey){
  names(desresWT[[listitem]]) <- mutnameKey[[listitem]]
}
```

Focusing at the moment on TilS_nochr3 subset, then PPC_nochr3.
For the interactive capturing/labeling of points, the plotMA call needs to be run from the console (not the notebook.) Followed by the idx-identify line in the console, and clicking on the points in the plot window.
(For clearer labels: capture points that need labeled in one direction (above, below, to the right/left) and plot each separately with multiple identify calls. n defines how many points are being picked up, and pos is where the label will go relative to the point.)

These are easier to work with indvidually, out of the list. (Just swapping the dataset and mutant specificers, and the titles in the plot call. Then modifying the points and labeling as I go. Collecting points saved as I go in a list. (Helps to reference the significant values tables for each comparison.) Saving these manually from the plot window




```{r}
# # Setup for collecting data, just in case? Can call in the notebook
# maPoints <- vector("list", 6)
# names(maPoints) <- nameKey
# maPoints <- lapply(setNames(nameKey, nameKey), function(listitem) {
#   indv <- vector("list", length(mutnameKey[[listitem]]))
#   names(indv) <- mutnameKey[[listitem]]
#   return(indv)
# })
```


```{r}
# Specify which LFC shrinkage and results sets
singleresLFC <- resLFCWT[["PPC_nochr3"]][["strain_T41T_vs_WT"]]
any(is.na(singleresLFC$log2FoldChange)) # General check (follow up with which() call.)
# Find min/max shrunken logfoldchange to adjust limits
min(singleresLFC$log2FoldChange, na.rm = TRUE)
max(singleresLFC$log2FoldChange, na.rm = TRUE)

# Get index number of specific locusTAg
match("Bcen2424_4447", rownames(singleresLFC))
match("Bcen2424_1362", rownames(singleresLFC))
```

```{r}
# Create list of genes of interest and set parameters for plot
hR <- 4333
hL <- 1321
highlight <- c(hR, hL)
ylow <- -1
yhigh <- 2
ptitle <- "T41T vs WT, PPC subset with Chromosomes 1&2"
```

```{r}
# Call these from console: Create plot, add two sets of texts to the right/left of points, highlight selected points
plotMA(singleresLFC, ylim=c(ylow,yhigh), main = ptitle)
text(singleresLFC$baseMean[hR], singleresLFC$log2FoldChange[hR], rownames(singleresLFC)[hR], pos = 4, cex = 1, offset = 0.5)
text(singleresLFC$baseMean[hL], singleresLFC$log2FoldChange[hL], rownames(singleresLFC)[hL], pos = 3, cex = 1, offset = 0.5)
points(singleresLFC$baseMean[highlight], singleresLFC$log2FoldChange[highlight], col = "blue")
```

To augment the plots, while the plot is still displayed in the plot pane:
*Decide how many points to select (grouped by what position to put the labels in), change n =
*Run one of the idx lines (character for position of label) from console and select points.
```{r}
# MUST CALL THESE FROM CONSOLE INDIVIDUALLY
# Identifying points by clicking, n is number to choose.
# To use the text() call to label, pos and plot must be false. To list index numbers as lables: plot=TRUE, pos is directional.
idxR <- identify(singleresLFC$baseMean, singleresLFC$log2FoldChange, n = 3, tolerance = 0.1, plot = FALSE, pos = FALSE)
text(singleresLFC$baseMean[idxR], singleresLFC$log2FoldChange[idxR],
     rownames(singleresLFC)[idxR], pos = 4, cex = 0.5, offset = 0.2)

idxL <- identify(singleresLFC$baseMean, singleresLFC$log2FoldChange, n = 1, tolerance = 0.1, plot = FALSE, pos = FALSE)
text(singleresLFC$baseMean[idxL], singleresLFC$log2FoldChange[idxL],
     rownames(singleresLFC)[idxL], pos = 2, cex = 0.5, offset = 0.2)

idxT <- identify(singleresLFC$baseMean, singleresLFC$log2FoldChange, n = 4, tolerance = 0.1, plot = FALSE, pos = FALSE)
text(singleresLFC$baseMean[idxT], singleresLFC$log2FoldChange[idxT],
     rownames(singleresLFC)[idxT], pos = 3, cex = 0.5, offset = 0.2)

idxB <- identify(singleresLFC$baseMean, singleresLFC$log2FoldChange, n = 1, tolerance = 0.1, plot = FALSE, pos = FALSE)
text(singleresLFC$baseMean[idxB], singleresLFC$log2FoldChange[idxB],
     rownames(singleresLFC)[idxB], pos = 1, cex = 0.5, offset = 0.2)


# Saving the points, but this closes the plot if run in the notebook (run in the console!)
holder <- rownames(singleresLFC)[idx] #Initial creation of list
holder <- c(holder, list(rownames(singleresLFC)[idx])) #Repeat this for every point-saving call
maPoints[["TilS_nochr3"]][["A244T"]] <- holder
```
Also grabbing points that are too close together to label, stored in other_upper/lower.

*These must all be saved from the plot pane.*


May make more sense to store all of the text to be plotted, then plot with textplot() from wordcloud package.
```{r}
idx <- identify(singleresLFC$baseMean, singleresLFC$log2FoldChange, n = 18, tolerance = 0.1, plot = FALSE, pos = FALSE)
textplot(singleresLFC$baseMean[idx], singleresLFC$log2FoldChange[idx], rownames(singleresLFC)[idx], cex=0.5, new=FALSE, show.lines = TRUE)
```
Not working, so sad :(

(In creating these, I found that there are null entires for these genes in this subset. So far only looked through the TilS_nochr3 and PPC_nochr3 subsets. Have not fully checked all sets. Gene name and index listed.)
```{r}
nulls <- vector("list", 6)
names(nulls) <- nameKey
nulls <- lapply(setNames(nameKey, nameKey), function(listitem) {
  indv <- vector("list", length(mutnameKey[[listitem]]))
  names(indv) <- mutnameKey[[listitem]]
  return(indv)
})
nulls[["PPC_nochr3"]][["S3C"]] <- which(is.na(resLFCWT[["PPC_nochr3"]][["strain_S3C_vs_WT"]]$log2FoldChange))
nulls[["PPC_nochr3"]][["T41T"]] <- which(is.na(resLFCWT[["PPC_nochr3"]][["strain_T41T_vs_WT"]]$log2FoldChange))
nulls
```

### MA-plots automated
Using ggpubr's ggmaplot to automate plot production, with more customizability in ggplot2.
This inital setup has the top 20 genes by shrunken log2fold change (other option is by padj.)
Highlighted gene is isocitrate lyase.
```{r}
falseDisRate <- 0.05
shrunklfc.cutoff <- 0.58 # Corresponds to 1.5
lfc.cutoff <- 1.5
highlight <- 4333 # Indeces of genes to highlight in plot
# Create the file structure
lapply(names(resLFCWT), function(listitem){
    dir.create(here::here("results", "MAplots", listitem), recursive = TRUE, showWarnings = FALSE)
})
maPlots <- lapply(setNames(names(resLFCWT), names(resLFCWT)), function(listitem){
  lapply(setNames(names(resLFCWT[[listitem]]), names(resLFCWT[[listitem]])), function(pairing){
    df.highlight <- as.data.frame(resLFCWT[[listitem]][[pairing]])[highlight, ]
    df.highlight$logMean <- log2(df.highlight$baseMean)
    t <- gsub("strain_", "", pairing, fixed = TRUE)
    t <- gsub("_vs_WT", "", t, fixed = TRUE)
    dset <- c("Full"="Full dataset, without plasmid",
              "Full_nochr3"="Full dataset, Chromosomes 1&2",
              "TilS"="TilS subset, without plasmid",
              "TilS_nochr3"="TilS subset, Chromosomes 1&2",
              "PPC"="PPC subset, without plasmid",
              "PPC_nochr3"="PPC subset, Chromosomes 1&2")
    p <- ggmaplot(resLFCWT[[listitem]][[pairing]], main = paste(t, "vs WT ",dset, sep=" "),
                    fdr = falseDisRate, fc = shrunklfc.cutoff, size = 0.4,
                    palette = c("#B31B21", "#1465AC", "darkgray"),
                    genenames = as.vector(rownames(resLFCWT[[listitem]][[pairing]])),
                    legend = "top", top = 20,
                    select.top.method = "fc", #Can select top by foldchange "fc" or "padj"
                    font.label =  11,
                    font.legend = "bold",
                    font.main = "bold",
                    ylab = "Shrunken log2fold change",
                    ggtheme = ggplot2::theme_minimal()) +
      geom_label_repel(data = df.highlight, aes(x=logMean, y=log2FoldChange),
                       label = rownames(df.highlight), fill = NA, xlim = c(NA, 6))
  ggsave(here::here("results", "MAplots", listitem, paste("MAplot", pairing, listitem, "eps", sep=".")),
         plot = p, width = 9, height = 7, units = "in", device = "eps")
  ggsave(here::here("results", "MAplots", listitem, paste("MAplot", pairing, listitem, "pdf", sep=".")),
         plot = p, width = 9, height = 7, units = "in", device = "pdf")
    return(p)
  })
})
```


The manual creation that the automated form is based off of.
```{r}
dir.create(here::here("results", "pairwise", "unshrunk_MAplots"))
ggsingleShrunk <- resLFCWT[["TilS_nochr3"]][["strain_A244T_vs_WT"]]
ggsingleRes <- desresWT[["TilS_nochr3"]][["A244T"]]

df.highlight <- as.data.frame(ggsingleShrunk)[highlight, ]
df.noShrink.highlight <- as.data.frame(ggsingleRes)[highlight, ]
df.highlight$logMean <- log2(df.highlight$baseMean)
df.noShrink.highlight$logMean <- log2(df.noShrink.highlight$baseMean)

pShrunk <- ggmaplot(ggsingleShrunk, main = "A244T to WT, Chromosomes 1&2, TilS Subset",
                    fdr = falseDisRate, fc = shrunklfc.cutoff, size = 0.4,
                    palette = c("#B31B21", "#1465AC", "darkgray"),
                    genenames = as.vector(rownames(ggsingleShrunk)),
                    legend = "top", top = 20,
                    select.top.method = "padj", #Can select top by foldchange "fc" or "padj"
                    font.label =  11,
                    font.legend = "bold",
                    font.main = "bold",
                    ylab = "Shrunken log2fold change",
                    ggtheme = ggplot2::theme_minimal()) +
#  geom_point(data = df.highlight, mapping = aes(x=logMean, y=log2FoldChange), color = "orange", size = 0.8) + 
  geom_label_repel(data = df.highlight, aes(x=logMean, y=log2FoldChange),
                   label = rownames(df.highlight), fill = NA, xlim = c(NA, 6))
print(pShrunk)
ggsave(here::here("results", "MAplots", "MA_A244TtoWT_TilS_nochr3.eps"),
       plot = pShrunk ,device = "eps")
# ggsave(here::here("results", "MAplots", "MA_A244TtoWT_TilS_nochr3.svg"),
#        plot = pShrunk ,device = "svg")
ggsave(here::here("results", "MAplots", "MA_A244TtoWT_TilS_nochr3.pdf"),
       plot = pShrunk ,device = "pdf")
```

Looking at the unshrunken (original) sets.
```{r}
plfc <- ggmaplot(ggsingleRes, main = "A244T to WT, Chromosomes 1&2, TilS Subset",
                 fdr = falseDisRate, fc = lfc.cutoff, size = 0.4,
                 palette = c("#B31B21", "#1465AC", "darkgray"),
                 genenames = as.vector(rownames(ggsingleRes)),
                 legend = "top", top = 20,
                 select.top.method = "fc",
                 font.label =  11,
                 font.legend = "bold",
                 font.main = "bold",
                 ylab = "Log2fold change",
                 ggtheme = ggplot2::theme_minimal()) +
  geom_label_repel(data = df.noShrink.highlight, aes(x=logMean, y=log2FoldChange),
                   label = rownames(df.noShrink.highlight), fill = NA, xlim = c(NA, 6))
print(plfc)

ggsave(here::here("results", "MAplots", "unshrunk_MAplots", "MA_unshrunk_A244TtoWT_TilS_nochr3.eps"),
       plot = plfc ,device = "eps")
# ggsave(here::here("results", "MAplots", "unshrunk_MAplots", "MA_unshrunk_A244TtoWT_TilS_nochr3.svg"),
#        plot = plfc ,device = "svg")
ggsave(here::here("results", "MAplots", "unshrunk_MAplots", "MA_unshrunk_A244TtoWT_TilS_nochr3.pdf"),
       plot = plfc ,device = "pdf")
```



# Gviz?
*Current goals:*

*Update pcaExplorer figures, re-sizing and formatting the pcaExplorer top/bottom loading figures
*Look at changing hypothesis/threshold for Wald test in comparisons to WT
*Back through pcaExplorer, look at gene tab
*BioCyc for pathway help
*Nail down regions of interest
*Attaching the GRanges data to the DESEqDataSets.
  +Do I need rtracklayer?
*gviz up and running, requires pairwise comparisons... and grange, and lfcShrink

(remove tRNA mutant from TilS set and see how things change? The ribosomal RNA appears to be dominating results. Should I remove only the ribosomal?)

```{r}
library(GenomicFeatures)
txdb <- makeTxDbFromGFF(here::here("data", "references", "GCA_000203955.1_ASM20395v1_genomic.gff"))
txdb
txdbRS <- makeTxDbFromGFF(here::here("data", "references", "GCF_000203955.1_ASM20395v1_genomic.gff")) # RefSeq?
txdbRS
```
It looks like the GenBank version doesn't include the RNA features, but the RefSeq does. cds nrow is different, too. RefSeq dropped: gene-BCEN2424_RS16455, gene-BCEN2424_RS17110, gene-BCEN2424_RS33785, gene-BCEN2424_RS34535OK


```{r}
sessionInfo()
```
